### PR TITLE
ADD new surrogate benchmark for the exploring openml paper

### PR DESCRIPTION
**Pull request by mfeurer**

### CODE REVIEW

**Comment by mlindauer** [comment]
Main problem:
```
module = importlib.import_module("hpolib.benchmarks.surrogates.exploring_openml")
benchmark = getattr(module, "XGBoost_3")(rng=seed)
```
Further issues:
 
- [x] meta_information is missing (in particular `num_function_evals`, `name`, `f_opt` (can be set to 0))
- [x] please remove the print statements

**Comment by mfeurer** [comment]
I just updated the code a bit, the output of:
```
import importlib

module = importlib.import_module("hpolib.benchmarks.surrogates.exploring_openml")
benchmark = getattr(module, "XGBoost_3")(rng=1)
print(benchmark)
```
is now:
```
/home/feurerm/miniconda/3-4.5.4/envs/smac3/bin/python3 /home/feurerm/sync_dir/projects/HPOlib2/HPOlib2/bug.py
/home/feurerm/miniconda/3-4.5.4/envs/smac3/lib/python3.6/site-packages/sklearn/externals/joblib/externals/cloudpickle/cloudpickle.py:47: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
/home/feurerm/miniconda/3-4.5.4/envs/smac3/lib/python3.6/site-packages/lightgbm/basic.py:1209: UserWarning: categorical_feature in Dataset is overridden.
New categorical_feature is [1]
  'New categorical_feature is {}'.format(sorted(list(categorical_feature))))
<hpolib.benchmarks.surrogates.exploring_openml.XGBoost_3 object at 0x7fd6f7f8cda0>

Process finished with exit code 0
```
It takes a while (~5 minutes) to build the surrogate, though.

**Comment by mlindauer** [comment]
Now, it takes really a lot of time. I'm trying to import XGBoost_3 now for roughly half an hour. Unfortunately, there is not output and hence, I don't know what it is doing at all. Could you please some (sparse) output such that a user does not believe that he/she hits an infinite loop?

Is the reproducibility still ensured since the HPO runs every time again? So, can I validate the configuration I found on my local machine again and it will return the same results?

File: hpolib/benchmarks/surrogates/exploring_openml.py [hunk 1]
```diff
@@ -0,0 +1,515 @@
...
+    """
+    url = None
+
+    def __init__(self, dataset_id, n_splits=10, rng=None):
+        """
+
+        Parameters
+        ----------
+        dataset_id: int
+            Dataset Id as given in Table 2.
```
Comment (KEggensperger) [inline hpolib/benchmarks/surrogates/exploring_openml.py]: Table 2? Which Table 2 ...
Could you maybe add a list of valid ids in the meta_data?

File: hpolib/benchmarks/surrogates/exploring_openml.py [hunk 2]
```diff
@@ -0,0 +1,515 @@
...
+        self.classifier = self.__class__.__name__.split('_')[0]
+
+        surrogate_dir = os.path.join(hpolib._config.data_dir, "ExploringOpenML", 'surrogates')
+        if not os.path.exists(surrogate_dir):
+            os.makedirs(surrogate_dir, exist_ok=True)
+        surrogate_file_name = os.path.join(
+            surrogate_dir,
+            'surrogate_%s_%d.pkl.gz' % (self.classifier, self.dataset_id),
+        )
+        if os.path.exists(surrogate_file_name):
```
Comment (KEggensperger) [inline hpolib/benchmarks/surrogates/exploring_openml.py]: If I run this benchmark once and then use n_splits=10 and I run it again with n_splits=5 will this update the surrogate?

File: hpolib/benchmarks/surrogates/exploring_openml.py [hunk 3]
```diff
@@ -0,0 +1,515 @@
...
+class ExploringOpenML(AbstractBenchmark):
+    """Surrogate benchmarks based on the data from Automatic Exploration of Machine Learning
+    Benchmarks on OpenML by Kühn et al..
+
+    Data is obtained from:
+    https://figshare.com/articles/OpenML_R_Bot_Benchmark_Data_final_subset_/5882230
+    """
+    url = None
+
+    def __init__(self, dataset_id, n_splits=10, rng=None):
```
Comment (KEggensperger) [inline hpolib/benchmarks/surrogates/exploring_openml.py]: Could you also add a rebuild flag? (related to my comment below)

File: hpolib/benchmarks/surrogates/exploring_openml.py [hunk 4]
```diff
@@ -0,0 +1,515 @@
...
+        min_samples_leaf = UniformIntegerHyperparameter('min_samples_leaf', 1, 20, log=True)
+        max_features = UniformFloatHyperparameter('max_features', 0.5, 1.0)
+        bootstrap = CategoricalHyperparameter('bootstrap', [True, False])
+        cs.add_hyperparameters([
+            min_samples_split,
+            min_samples_leaf,
+            max_features,
+            bootstrap,
+        ])
+        cs.seed(1)
```
Comment (KEggensperger) [inline hpolib/benchmarks/surrogates/exploring_openml.py:221]: This makes constructing the surrogate deterministic, correct?

File: hpolib/benchmarks/surrogates/exploring_openml.py [hunk 5]
```diff
@@ -0,0 +1,515 @@
...
+        highest_correlations = -np.inf
+        highest_correlations_by_fold = np.array((n_splits,)) * -np.inf
+        best_config = cs.get_default_configuration()
+        n_iterations = 0
+        while True:
+            self.logger.debug('Random search iteration %d/%d.', n_iterations, 40)
+            n_iterations += 1
+
+            # Stopping condititions
+            if n_iterations > 40:
```
Comment (KEggensperger) [inline hpolib/benchmarks/surrogates/exploring_openml.py]: This should be an input with a default to the benchmark.

File: hpolib/benchmarks/surrogates/exploring_openml.py [hunk 6]
```diff
@@ -0,0 +1,515 @@
...
+                self.logger.debug('Stopping random search due to good performance!')
+                break
+            elif highest_correlations > 0.95 and n_iterations > 20:
+                self.logger.debug('Stopping random search due to good performance!')
+                break
+            elif highest_correlations > 0.99 and n_iterations > 10:
+                self.logger.debug('Stopping random search due to good performance!')
+                break
+            elif highest_correlations > 0.999:
+                self.logger.debug('Stopping random search due to good performance!')
```
Comment (KEggensperger) [inline hpolib/benchmarks/surrogates/exploring_openml.py]: Where do these numbers come from?

File: hpolib/benchmarks/surrogates/exploring_openml.py [hunk 7]
```diff
@@ -0,0 +1,515 @@
...
+            nan_rows = ~np.isfinite(features[:, i])
+            features[nan_rows, i] = hp.normalized_default_value
+
+    @AbstractBenchmark._check_configuration
+    def objective_function(self, x, **kwargs):
+        x = x.get_array().reshape((1, -1))
+        self.impute_with_defaults(x)
+        y = self.regressor.predict(x)
+        y = y[0]
+        return {'function_value': y}
```
Comment (KEggensperger) [inline hpolib/benchmarks/surrogates/exploring_openml.py]: Is there no runtime we can return?

File: hpolib/benchmarks/surrogates/exploring_openml.py [hunk 8]
```diff
@@ -0,0 +1,515 @@
...
+    @AbstractBenchmark._check_configuration
+    def objective_function_test(self, x, **kwargs):
+        raise NotImplementedError()
+
+    @staticmethod
+    def get_meta_information():
+        return {
+            'num_function_evals': 50,
+            'name': 'Exploring_OpenML',
+            'f_opt': 0,
```
Comment (KEggensperger) [inline hpolib/benchmarks/surrogates/exploring_openml.py]: Are you sure about this?

File: hpolib/benchmarks/surrogates/exploring_openml.py [hunk 9]
```diff
@@ -0,0 +1,515 @@
...
+    for dataset_id_ in all_datasets:
+        benchmark_string = """class %s_%d(%s):
+         def __init__(self, n_splits=10, rng=None):
+             super().__init__(dataset_id=%d, n_splits=n_splits, rng=rng)
+    """ % (model_class.__name__, dataset_id_, model_class.__name__, dataset_id_)
+
+        exec(benchmark_string)
+
+
+if __name__ == '__main__':
```
Comment (KEggensperger) [inline hpolib/benchmarks/surrogates/exploring_openml.py:601]: Why is this here?

File: hpolib/benchmarks/surrogates/exploring_openml.py (hunk 1 above)
Comment (KEggensperger) [reply hpolib/benchmarks/surrogates/exploring_openml.py]: Okay, I just found the list. So this is then rather an Abstract class that should not be used directly. Could you add that somewhere in the comments?

**Review by KEggensperger — CHANGES_REQUESTED** [review]
All in all this looks good, but I do have some minor comments?

File: hpolib/benchmarks/surrogates/exploring_openml.py (hunk 1 above)
Comment (mfeurer) [reply hpolib/benchmarks/surrogates/exploring_openml.py]: Added this.

File: hpolib/benchmarks/surrogates/exploring_openml.py (hunk 2 above)
Comment (mfeurer) [reply hpolib/benchmarks/surrogates/exploring_openml.py]: No.

File: hpolib/benchmarks/surrogates/exploring_openml.py (hunk 3 above)
Comment (mfeurer) [reply hpolib/benchmarks/surrogates/exploring_openml.py]: Yes, I just added this.

File: hpolib/benchmarks/surrogates/exploring_openml.py (hunk 4 above)
Comment (mfeurer) [reply hpolib/benchmarks/surrogates/exploring_openml.py:221]: Correct.

File: hpolib/benchmarks/surrogates/exploring_openml.py (hunk 5 above)
Comment (mfeurer) [reply hpolib/benchmarks/surrogates/exploring_openml.py]: I added this, and also use it in the unit tests now.

File: hpolib/benchmarks/surrogates/exploring_openml.py (hunk 6 above)
Comment (mfeurer) [reply hpolib/benchmarks/surrogates/exploring_openml.py]: I removed these numbers. These were based on intuition and the results your AAAI paper.

File: hpolib/benchmarks/surrogates/exploring_openml.py (hunk 7 above)
Comment (mfeurer) [reply hpolib/benchmarks/surrogates/exploring_openml.py]: I just added the runtime.

File: hpolib/benchmarks/surrogates/exploring_openml.py (hunk 9 above)
Comment (mfeurer) [reply hpolib/benchmarks/surrogates/exploring_openml.py:601]: For some testing if all methods can be instantiated.

File: hpolib/benchmarks/surrogates/exploring_openml.py (hunk 8 above)
Comment (mfeurer) [reply hpolib/benchmarks/surrogates/exploring_openml.py]: I'd be happy to change this, but given the fact that the class is abstract I can't add a way to access the data and therefore need to assume the lowest possible value. I added two functions to retrieve the empirical min and max which is necessary to compute some metrics.

**Comment by mfeurer** [comment]
Okay @KEggensperger I think I addressed all your requests.

**Review by KEggensperger — APPROVED** [review]

**Pull request merged**